jsonschema/json.go

```diff
@@ -53,3 +58,90 @@ func (d Definition) MarshalJSON() ([]byte, error) {
 		Alias: (Alias)(d),
 	})
 }
+
+type SchemaWrapper[T any] struct {
+	data   T
+	schema Definition
+}
+
+func (r SchemaWrapper[T]) Schema() Definition {
+	return r.schema
+}
+
+func (r SchemaWrapper[T]) MarshalJSON() ([]byte, error) {
+	return json.Marshal(r.schema)
+}
+
+func (r SchemaWrapper[T]) Unmarshal(content string) (*T, error) {
+	var v T
+	err := Unmarshal(r.schema, []byte(content), &v)
+	if err != nil {
+		return nil, err
+	}
+	return &v, nil
+}
+
+func (r SchemaWrapper[T]) String() string {
+	bytes, _ := json.MarshalIndent(r.schema, "", "  ")
+	return string(bytes)
+}
+
+func Warp[T any](v T) SchemaWrapper[T] {
+	return SchemaWrapper[T]{
+		data:   v,
+		schema: reflectSchema(reflect.TypeOf(v)),
+	}
+}
+
+func reflectSchema(t reflect.Type) Definition {
+	var d Definition
+	switch t.Kind() {
+	case reflect.String:
+		d.Type = String
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
+		reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		d.Type = Integer
+	case reflect.Float32, reflect.Float64:
+		d.Type = Number
+	case reflect.Bool:
+		d.Type = Boolean
+	case reflect.Slice, reflect.Array:
+		d.Type = Array
+		items := reflectSchema(t.Elem())
+		d.Items = &items
+	case reflect.Struct:
+		d.Type = Object
+		d.AdditionalProperties = false
+		properties := make(map[string]Definition)
+		var requiredFields []string
+		for i := 0; i < t.NumField(); i++ {
+			field := t.Field(i)
+			jsonTag := field.Tag.Get("json")
+			var required = true
+			if jsonTag == "" {
+				jsonTag = field.Name
+			} else if strings.HasSuffix(jsonTag, ",omitempty") {
+				jsonTag = strings.TrimSuffix(jsonTag, ",omitempty")
+				required = false
+			}
+
+			item := reflectSchema(field.Type)
+			description := field.Tag.Get("description")
+			if description != "" {
+				item.Description = description
+			}
+			properties[jsonTag] = item
+
+			if s := field.Tag.Get("required"); s != "" {
+				required, _ = strconv.ParseBool(s)
+			}
+			if required {
+				requiredFields = append(requiredFields, jsonTag)
+			}
+		}
+		d.Required = requiredFields
+		d.Properties = properties
+	default:
```

Comment:
I suggest we either return an error or panic here

Reply:
Thank you for the suggestion!